freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java

```diff
@@ -0,0 +1,185 @@
+import org.springframework.statemachine.action.Action;
+
+import com.sequenceiq.flow.core.Flow;
+import com.sequenceiq.flow.core.FlowParameters;
+import com.sequenceiq.freeipa.flow.stack.StackEvent;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.UpgradeCcmService;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.UpgradeCcmState;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.event.UpgradeCcmEvent;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.event.UpgradeCcmFailureEvent;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.event.UpgradeCcmTriggerEvent;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.selector.UpgradeCcmStateSelector;
+
+@Configuration
+public class UpgradeCcmActions {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UpgradeCcmActions.class);
+
+    @Inject
+    private UpgradeCcmService upgradeCcmService;
+
+    @Bean(name = UPGRADE_CCM_CHECK_PREREQUISITES_STATE_NAME)
+    public Action<?, ?> checkPrerequisites() {
+        return new AbstractUpgradeCcmAction<>(UpgradeCcmTriggerEvent.class) {
+            @Override
+            protected void doExecute(UpgradeCcmContext context, UpgradeCcmTriggerEvent payload, Map<Object, Object> variables) {
+                setOperationId(variables, payload.getOperationId());
+                setFinalChain(variables, payload.isFinal());
+                setChainedAction(variables, payload.isChained());
+                LOGGER.info("Starting checking prerequisites for CCM upgrade {}", payload);
+                upgradeCcmService.checkPrerequisitesState(context.getStack().getId());
+                sendEvent(context, UPGRADE_CCM_CHECK_PREREQUISITES_EVENT.create(context.getStack().getId()));
+            }
+        };
+    }
+
+    @Bean(name = UPGRADE_CCM_CHANGE_TUNNEL_STATE_NAME)
+    public Action<?, ?> changeTunnel() {
+        return new AbstractUpgradeCcmEventAction() {
+            @Override
+            protected void doExecute(UpgradeCcmContext context, UpgradeCcmEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Starting checking prerequisites for CCM upgrade {}", payload);
+                upgradeCcmService.changeTunnelState(context.getStack().getId());
+                sendEvent(context, UPGRADE_CCM_CHANGE_TUNNEL_EVENT.createBasedOn(payload));
+            }
+        };
+    }
+
+    @Bean(name = UPGRADE_CCM_PUSH_SALT_STATES_STATE_NAME)
+    public Action<?, ?> pushSaltStates() {
+        return new AbstractUpgradeCcmEventAction() {
+            @Override
+            protected void doExecute(UpgradeCcmContext context, UpgradeCcmEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Starting pushing salt states for CCM upgrade {}", payload);
+                upgradeCcmService.pushSaltStatesState(context.getStack().getId());
+                sendEvent(context, UPGRADE_CCM_PUSH_SALT_STATES_EVENT.createBasedOn(payload));
+            }
+        };
+    }
```

Comment:
I'm a bit confused what is the flow of this process is, maybe it's related to the fact that there is no description to the commit which has the expected logic details.
So what is the plane to make the changes on the host? Are you planning a full salt state upgrade which would first push down CCMV2 related configs and start it, then would come the CP rereg+healtcheck, then once again a salt state upgrade to disable old CCM.
Or are you planning to play with pushing down single states/pillars?

Reply:
pillar update (with top.sls) and then bootstrap salt (to reload pillar with updated top.sls and propagate new states)

Reply:
in that case it seems this should be a flowchain and the salt state update flow should be called

Reply:
Unfortunately the salt update flow is not prepared to receive payload from external object in OrchestratorConfigHandler
on the other hand, this flow also installs FreeIPA which is not desired in this case.
It would mean a bit more hack than the currently created customized flow in this PR.

Reply:
I'm still confused, the salt stuff should be idempotent, you should be able to run highstate at any time on a machine or the whole cluster and it should be fine.
Anything which is salt level logic should be controlled by pillars and grains. So you should be able to push down all the necessary states and pillars for the ccm upgrade using the flow. Yes, this will invoke a highstate, but that shouldn't be a problem at all, you might set the necessary grains before and everything should be in a good state after the highstate. Or you can add the grains later and call a state or highstate after that.

>  installs FreeIPA

I don't get it, FreeIPA is already installed, it won't be reinstalled (at least it shouldn't be reinstalled, if it is, then it's a bug)